File: open_raccoon/apps/raccoon_pages/lib/raccoon_pages/page.ex
defmodule RaccoonPages.Page do
  @moduledoc """
  Page schema for user-created web pages deployed to raccoon.page.
  """
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "pages" do
    belongs_to :creator, RaccoonAccounts.User
    belongs_to :agent, RaccoonAgents.Agent
    belongs_to :conversation, RaccoonChat.Conversation
    belongs_to :forked_from_page, RaccoonPages.Page, foreign_key: :forked_from

    field :title, :string
    field :slug, :string
    field :description, :string
    field :thumbnail_url, :string
    field :r2_path, :string
    field :deploy_url, :string
    field :custom_domain, :string
    field :version, :integer, default: 1
    field :visibility, Ecto.Enum, values: [:public, :unlisted, :private], default: :public
    field :view_count, :integer, default: 0

    has_many :versions, RaccoonPages.PageVersion

    timestamps(type: :utc_datetime_usec)
  end

  def changeset(page, attrs) do
    page
    |> cast(attrs, [
      :creator_id, :agent_id, :conversation_id, :title, :slug, :description,
      :thumbnail_url, :r2_path, :deploy_url, :custom_domain, :version,
      :forked_from, :visibility
    ])
    |> validate_required([:creator_id, :title, :slug, :r2_path])
    |> validate_length(:title, max: 255)
    |> validate_length(:slug, max: 128)
    |> validate_number(:version, greater_than: 0)
    |> validate_inclusion(:visibility, [:public, :unlisted, :private])
    |> unique_constraint([:creator_id, :slug])
    |> foreign_key_constraint(:creator_id)
    |> foreign_key_constraint(:agent_id)
    |> foreign_key_constraint(:conversation_id)
    |> foreign_key_constraint(:forked_from)
  end
end

File: WaiAgents/Sources/WaiAgentsKit/Models/Page.swift
import Foundation

public struct Page: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let creatorID: String
    public var agentID: String?
    public var conversationID: String?
    public var title: String
    public var slug: String
    public var description: String?
    public var thumbnailURL: URL?
    public var r2Path: String
    public var deployURL: URL?
    public var customDomain: String?
    public var version: Int
    public var forkedFrom: String?
    public var visibility: String
    public var viewCount: Int
    public let createdAt: Date
    public var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case creatorID = "creatorId"
        case agentID = "agentId"
        case conversationID = "conversationId"
        case title
        case slug
        case description
        case thumbnailURL = "thumbnailUrl"
        case r2Path
        case deployURL = "deployUrl"
        case customDomain
        case version
        case forkedFrom
        case visibility
        case viewCount
        case createdAt
        case updatedAt
    }

    public init(
        id: String,
        creatorID: String,
        agentID: String? = nil,
        conversationID: String? = nil,
        title: String,
        slug: String,
        description: String? = nil,
        thumbnailURL: URL? = nil,
        r2Path: String,
        deployURL: URL? = nil,
        customDomain: String? = nil,
        version: Int = 1,
        forkedFrom: String? = nil,
        visibility: String = "public",
        viewCount: Int = 0,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.creatorID = creatorID
        self.agentID = agentID
        self.conversationID = conversationID
        self.title = title
        self.slug = slug
        self.description = description
        self.thumbnailURL = thumbnailURL
        self.r2Path = r2Path
        self.deployURL = deployURL
        self.customDomain = customDomain
        self.version = version
        self.forkedFrom = forkedFrom
        self.visibility = visibility
        self.viewCount = viewCount
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

public struct PageVersion: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let pageID: String
    public let version: Int
    public let r2Path: String
    public var changes: String?
    public let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case pageID = "pageId"
        case version
        case r2Path
        case changes
        case createdAt
    }

    public init(
        id: String,
        pageID: String,
        version: Int,
        r2Path: String,
        changes: String? = nil,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.pageID = pageID
        self.version = version
        self.r2Path = r2Path
        self.changes = changes
        self.createdAt = createdAt
    }
}
